### PR TITLE
sql: improve known limitations for tables

### PR DESCRIPTION
**Pull request by benesch**
`INSERT ... SELECT` is now supported; all of `INSERT ... SELECT`, `UPDATE`, and `DELETE` have performance caveats that need to be called out.

@jkosh44 @mjibson — I need your help wording the performance limitation for `INSERT ... SELECT`, `UPDATE`, and `DELETE`. I'm not sure I got it right.

@jseldess — this whole thing could use a copyedit. Would love to hand this off to the docs team!

### Motivation

  * This PR fixes a bug reported on Slack.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

**Review by jseldess — COMMENTED** [review]
@benesch, this already looks very clean. Thank you! I'll help incorporate input from Josh and Matt, when that comes in.

One request: Can you help me track down existing GH issues for these limitations? I think we should link to issues, where we can.

File: doc/user/content/sql/delete.md [hunk 1]
```diff
@@ -21,7 +21,12 @@ _alias_ | Only permit references to _table_name_ as _alias_.
 
 ## Details
 
-`DELETE` cannot currently be used inside [transactions](../begin).
+### Known limitations
+
+* `DELETE` cannot be used inside [transactions](../begin).
+
+* **Low performance.** While processing a `DELETE` statement, Materialize cannot
+  process other `INSERT`, `UPDATE`, or `DELETE` statements.
```
Comment (jkosh44) [inline doc/user/content/sql/delete.md:29]: I think the one caveat to add to this (and all others), is that it's only true when the `DELETE` involves a read. That's almost all deletes so it's not a big deal, but maybe worth calling out.

**Comment by mjibson** [comment]
I think the new text is still a bit incorrect. `UPDATE`, `DELETE`, and `INSERT .. SELECT` can all reference other *tables*, but not sources:

```
materialize=> create table t1(i int);
CREATE TABLE
materialize=> create table t2(j int);
CREATE TABLE
materialize=> CREATE SOURCE counter
  FROM LOAD GENERATOR COUNTER
  (TICK INTERVAL '500ms')
  WITH (SIZE = '1');
CREATE SOURCE
materialize=> insert into t1 select * from t2;
INSERT 0 0
materialize=> insert into t1 select * from counter;
ERROR:  invalid selection: operation may only refer to user-defined tables
materialize=> delete from t1 where i in (select * from t2);
DELETE 0
materialize=> delete from t1 where i in (select * from counter);
ERROR:  invalid selection: operation may only refer to user-defined tables
materialize=> update t1 set i=1 where i in (select * from t2);
UPDATE 0
materialize=> update t1 set i=1 where i in (select * from counter);
ERROR:  invalid selection: operation may only refer to user-defined tables
```

**Comment by benesch** [comment]
> One request: Can you help me track down existing GH issues for these limitations? I think we should link to issues, where we can.

I don't think we have existing issues for any of these limitations! Adapter folks: can you fact check me on this, and possibly file new issues if the existing issues do not exist?

File: doc/user/content/sql/delete.md (hunk 1 above)
Comment (jseldess) [reply doc/user/content/sql/delete.md:29]: Sorry for the delay on this. @jkosh44, I don't understand what you mean by a delete involving a read. Can you explain that further?

**Comment by jseldess** [comment]
Pushed a commit to address @mjibson's clarification.

File: doc/user/content/sql/delete.md (hunk 1 above)
Comment (benesch) [reply doc/user/content/sql/delete.md:29]: I'm also confused. I think all `DELETE`s today involve a read?

Comment (jkosh44) [reply doc/user/content/sql/delete.md:29]: Yes, you're right, I was confusing myself. You can ignore this.

**Review by jseldess — APPROVED** [review]

File: doc/user/content/sql/delete.md (hunk 1 above)
Comment (benesch) [reply doc/user/content/sql/delete.md:29]: Realized we didn't have an issue for `RETRACT`/`DELYEET`, so I filed one: https://github.com/MaterializeInc/materialize/issues/19071

**Comment by jseldess** [comment]
Merging, now that there are no open questions.

**Pull request merged**